File: LeetCode/Medium/0814-binary-tree-pruning/0814-binary-tree-pruning.cpp
class Solution {
public:

    int fun( TreeNode* root ){

        if( root == nullptr ) return 0 ;

        int left = fun( root->left  );
        int right = fun( root->right );

        if(  left == 0  ){
            root->left = nullptr;
        }
        if(  right  == 0  ){
            root->right = nullptr;
        }
        return root->val + left + right  ;

    }

    TreeNode* pruneTree(TreeNode* root) 
    {
        int t = fun(root);
        if( t == 0  )
            return nullptr ;
        return root;    
    }
};

File: LeetCode/Medium/0814-binary-tree-pruning/0814-binary-tree-pruning.java

class Solution {
    int fun( TreeNode root ){
        if( root == null )
            return 0 ;
        int left = fun( root.left );
        int right = fun( root.right );

        if( left == 0 ) root.left = null ;
        if( right == 0 ) root.right = null ;
        return root.val + left + right ;
    }
    public TreeNode pruneTree(TreeNode root) {
        int t = fun( root );
        if( t == 0 ) return null;
        return root;
    }
}